fix(ivy): destroy injector when module is destroyed

crisbeto opened the pull request:
Destroys the module's injector when an `NgModule` is destroyed which in turn calls the `ngOnDestroy` methods on the instantiated providers.

This PR resolves FW-739.

mary-poppins commented:
You can preview e47e586 at https://pr27793-e47e586.ngbuilds.io/.

mary-poppins commented:
You can preview 27b8357 at https://pr27793-27b8357.ngbuilds.io/.

mhevery commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -101,7 +101,8 @@ export class R3Injector {
    * Flag indicating that this injector was previously destroyed.
    */
-  private destroyed = false;
+  get destroyed(): boolean { return this._destroyed; }
```
Why turning this into getter?

crisbeto replied:
I want to prevent people from being able to set it from the outside. It could also be turned into a `isDestroyed` method which'll generate slightly less ES5 code.

mary-poppins commented:
You can preview 5c60d61 at https://pr27793-5c60d61.ngbuilds.io/.

mary-poppins commented:
You can preview 93eac72 at https://pr27793-93eac72.ngbuilds.io/.

mary-poppins commented:
You can preview e2487fc at https://pr27793-e2487fc.ngbuilds.io/.

petebacondarwin commented on packages/elements/test/slots_spec.ts:
```diff
@@ -39,7 +39,10 @@ if (browserDetection.supportsCustomElements && browserDetection.supportsShadowDo
+
+      if (testContainer.parentNode) {
+        testContainer.parentNode.removeChild(testContainer);
+      }
```
Why this change? Does calling `testContainer.remove()` not remove it from its parent? That would seem like an obvious thing for such a method to do...

mary-poppins commented:
You can preview 82aca8c at https://pr27793-82aca8c.ngbuilds.io/.

mary-poppins commented:
You can preview 59be577 at https://pr27793-59be577.ngbuilds.io/.

mary-poppins commented:
You can preview 139c3b2 at https://pr27793-139c3b2.ngbuilds.io/.

kara commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -456,6 +457,7 @@ function removeListeners(lView: LView): void {
       }
     }
     lView[CLEANUP] = null;
+    tView.cleanup = null;
```
Hmm this doesn't look quite right since `TViews` are shared. If you remove one view instance, wouldn't this mean that when other view instances are removed, the cleanup fns would be gone?  Let's add a test for this ?

mary-poppins commented:
You can preview 434e8f8 at https://pr27793-434e8f8.ngbuilds.io/.

crisbeto committed:
fix(ivy): destroy injector when module is destroyed

Destroys the module's injector when an `NgModule` is destroyed which in turn calls the `ngOnDestroy` methods on the instantiated providers.

This PR resolves FW-739.

mary-poppins commented:
You can preview 1eebc40 at https://pr27793-1eebc40.ngbuilds.io/.

kara reviewed (approved):
LGTM

kara commented:
[presubmit](http://test/OCL:229851752:BASE:229851780:1547776542391:8b4e3ccb)

alxhub commented:
Considering @kara and @mhevery's approval to be global.
